### PR TITLE
nixos/nh: init

### PR DESCRIPTION
**Pull request by viperML**
## Description of changes

Mostly a copy of the upstream module: https://github.com/viperML/nh/blob/master/module.nix

Marking as draft to get reactions about it. I need to add the release notes.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by viperML** [comment]
Idk why the bot adds the `has module (update)` tag

File: nixos/modules/programs/nh.nix [hunk 1]
```diff
@@ -0,0 +1,93 @@
+{ config
+, lib
+, pkgs
+, ...
+}: {
+  meta.maintainers = [ lib.maintainers.viperML ];
+
+  options.programs.nh = {
+    enable = lib.mkEnableOption (lib.mdDoc "Yet another nix cli helper.");
```
Comment (drupol) [inline nixos/modules/programs/nh.nix:9]:
```suggestion
    enable = lib.mkEnableOption (lib.mdDoc "Yet another Nix CLI helper.");
```

File: nixos/modules/programs/nh.nix [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      systemPackages = [ config.programs.nh.package ];
+      variables = lib.mkIf (config.programs.nh.flake != null) {
+        FLAKE = config.programs.nh.flake;
+      };
+    };
+
+    systemd = lib.mkIf config.programs.nh.clean.enable {
+      services.nh-clean = {
+        description = "Nh clean";
+        script = "exec ${config.nh.package}/bin/nh clean all ${config.nh.clean.extraArgs}";
```
Comment (drupol) [inline nixos/modules/programs/nh.nix:80]:
```suggestion
        script = "exec ${lib.getExe config.nh.package} clean all ${config.nh.clean.extraArgs}";
```

**Review by drupol — DISMISSED** [review]

File: nixos/modules/programs/nh.nix [hunk 3]
```diff
@@ -0,0 +1,93 @@
+{ config
+, lib
+, pkgs
+, ...
+}: {
+  meta.maintainers = [ lib.maintainers.viperML ];
+
+  options.programs.nh = {
+    enable = lib.mkEnableOption (lib.mdDoc "Yet another Nix CLI helper.");
```
Comment (eclairevoyant) [inline nixos/modules/programs/nh.nix:9]: `lib.mdDoc` is a no-op

File: nixos/modules/programs/nh.nix [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      }
+
+      {
+        assertion = (config.nh.flake != null) -> !(lib.hasSuffix ".nix" config.nh.flake);
+        message = "nh.flake must be a directory, not a nix file";
+      }
+
+      # Not strictly required, but the user probably doesn't want 2 gc services
+      {
+        assertion = config.programs.nh.clean.enable -> !config.nix.gc.automatic;
```
Comment (eclairevoyant) [inline nixos/modules/programs/nh.nix:65]: better as a warning?

File: nixos/modules/programs/nh.nix (hunk 3 above)
Comment (eclairevoyant) [reply nixos/modules/programs/nh.nix:9]: also we should remove the trailing period when using `mkEnableOption`

Comment (viperML) [reply nixos/modules/programs/nh.nix:9]: Shouldn't all module descriptions include mdDoc ?

**Review by drupol — COMMENTED** [review]
Hi mate,

I'm not that old yet to review NixOS module correctly. I need more XP. I'll leave this to one of the other maintainer.

**Review by eclairevoyant — COMMENTED** [review]
From my view:
* Usually in program modules, there's a let-binding at the top with
```nix
let
  cfg = config.programs.nh;
in
```
which helps avoid redundancy / improve readability
* The other `mdDoc` usages can be removed as well

Otherwise seems ok to me

**Comment by viperML** [comment]
> The other mdDoc usages can be removed as well

All of them?

File: nixos/modules/programs/nh.nix [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        message = "programs.nh.clean.enable requires programs.nh.enable";
+      }
+
+      {
+        assertion = (cfg.flake != null) -> !(lib.hasSuffix ".nix" cfg.flake);
+        message = "nh.flake must be a directory, not a nix file";
+      }
+    ];
+
+    environment = lib.mkIf (cfg.enable) {
```
Comment (yunfachi) [inline nixos/modules/programs/nh.nix:73]:
```suggestion
    environment = lib.mkIf cfg.enable {
```
`(cfg.enable)` useless parentheses

File: nixos/modules/programs/nh.nix [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+          See `nh clean all --help` for more information.
+        '';
+      };
+    };
+  };
+
+  config = {
+    warnings =
+      if (!(cfg.clean.enable -> !config.nix.gc.automatic)) then [
+        "programs.nh.clean.enable and nix.gc.automatic are both enabled. Please use one or the other to avoid conflict."
```
Comment (SuperSandro2000) [inline nixos/modules/programs/nh.nix:57]:
```suggestion
        "programs.nh.clean.enable and nix.gc.automatic conflict with each other, please decide for one of them."
```

File: nixos/modules/programs/nh.nix [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+          See `nh clean all --help` for more information.
+        '';
+      };
+    };
+  };
+
+  config = {
+    warnings =
+      if (!(cfg.clean.enable -> !config.nix.gc.automatic)) then [
```
Comment (SuperSandro2000) [inline nixos/modules/programs/nh.nix:56]: Shouldn't this be an assertion?

File: nixos/modules/programs/nh.nix [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        "programs.nh.clean.enable and nix.gc.automatic are both enabled. Please use one or the other to avoid conflict."
+      ] else [ ];
+
+    assertions = [
+      # Not strictly required but probably a good assertion to have
+      {
+        assertion = cfg.clean.enable -> cfg.enable;
+        message = "programs.nh.clean.enable requires programs.nh.enable";
+      }
```
Comment (SuperSandro2000) [inline nixos/modules/programs/nh.nix:65]: Why not set this option to true?

File: nixos/modules/programs/nh.nix (hunk 7 above)
Comment (viperML) [reply nixos/modules/programs/nh.nix:56]: Was suggested by @eclairevoyant before: https://github.com/NixOS/nixpkgs/pull/294923#discussion_r1519365725

File: nixos/modules/programs/nh.nix (hunk 6 above)
Comment (eclairevoyant) [reply nixos/modules/programs/nh.nix:57]: This is less correct than the original; what was the intention behind this suggestion?

File: nixos/modules/programs/nh.nix (hunk 7 above)
Comment (eclairevoyant) [reply nixos/modules/programs/nh.nix:56]: I don't think we need to arbitrarily block settings if they don't _actually_ conflict.

**Review by eclairevoyant — CHANGES_REQUESTED** [review]
Seems fine otherwise

**Review by eclairevoyant — APPROVED** [review]

**Comment by viperML** [comment]
thanks

**Pull request merged**